test/apps/binary-search.c

```diff
@@ -0,0 +1,82 @@
+/*
+ *  Copyright (c) 2016 Intel Corporation. All rights reserved.
+ *  This software is available to you under the BSD license below:
+ *
+ *      Redistribution and use in source and binary forms, with or
+ *      without modification, are permitted provided that the following
+ *      conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <shmem.h>
+
+#define N_PER_PE 10
+
+/* N_PER_PE*N_PES sorted, distributed shared array */
+int keys[N_PER_PE];
+
+static int binary_search(int key) {
+    int low, mid, high;
+
+    low = 0;
+    high = shmem_n_pes()*N_PER_PE;
+
+    while(low < high) {
+        int val;
+
+        mid = low + (high-low)/2;
+        val = shmem_int_g(&keys[mid%N_PER_PE], mid/N_PER_PE);
+
+        if(val == key) {
+            return mid;
+        } else if(val < key) {
+            low = mid;
+        } else {
+            high = mid;
+        }
+    }
+
+    return -1;
+}
+
+int main(int argc, char **argv) {
+    int i, errors = 0;
+
+    shmem_init();
+
+    for (i = 0; i < N_PER_PE; i++)
+        keys[i] = N_PER_PE * shmem_my_pe() + i;
+
+    shmem_barrier_all();
+
+    for (i = 0; i < N_PER_PE * shmem_n_pes(); i++) {
```

Comment:
This would be a pretty good stress-test for evaluating a fabric. It might be a bit better for that if the loop is offset by shmem_my_pe()*N_PER_PE so it doesn't bottleneck on a particular node.

Aside from that, it looks correct to me.